publish workaround for RWM volumes becoming inaccessible after node daemonset pod restarts

divyenpatel opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is publishing a workaround customers need to perform for RWM volume before upgrading CSI Driver.
Refer to https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/1216 for more detail.

Preview link - https://deploy-preview-1220--kubernetes-sigs-vsphere-csi-driver.netlify.app/driver-deployment/upgrade.html

**Special notes for your reviewer**:
Automation for the steps mentioned in the workaround will be published later on this repository. We will update the documentation at that time.


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
publish workaround for RWM volumes becoming inaccessible after node daemonset pod restarts
```

xing-yang commented:
/approve

SandeepPissay commented on docs/book/driver-deployment/upgrade.md:
```diff
@@ -7,9 +7,57 @@ Before upgrading driver, please make sure to keep roles and privileges up to dat
+
+After all vSAN file share volumes are remounted on the worker VMs, you can proceed with upgrading the driver.
+
+We are in the process of automating the above steps so until we have automation available we have to execute the workaround step mentioned above manually.
```
I'm curious to know what is the automation we are mentioning here. 

So, after upgrading the CSI driver to the latest, why not ask the customer to delete the pod and let Kubernetes recreate it. Of course, I'm assuming that they have a Deployment or StatefulSet managing the pod here and there is no standalone Pod.

divyenpatel replied:
> So, after upgrading the CSI driver to the latest, why not ask the customer to delete the pod and let Kubernetes recreate it.

If we delete the pod, pod will be recreated with the same old driver which does not hostnetwork set to true.

and for frozen volumes, when we delete the pod, the pod remains in the terminating state. That's the reason we are providing a workaround of remounting volume from Guest VM.


> I'm curious to know what is the automation we are mentioning here.

we are thinking about providing a container image to fix all such volumes to help remount on the node VM. In addition to automation, it will help eliminate the requirement of installing `nfs-common` utils on worker VMs.

shalini-b commented on docs/book/driver-deployment/upgrade.md:
```diff
@@ -4,14 +4,82 @@
 
-For example, if you want to upgrade driver from `v2.0.0` release to `v2.0.1` release deployed on the Kubernetes Cluster installed on `vSphere 67u3` release
+To upgrade driver from `v2.0.0` release to `v2.0.1` release deployed on the Kubernetes Cluster installed on `vSphere 67u3` release
 
```
Can we mention that this is an example of upgrading driver from `2.0.0` to `2.3.0`? 
`v2.0.1` is incorrect in this sentence.

divyenpatel replied:
fixed.

shalini-b commented on docs/book/driver-deployment/upgrade.md:
````diff
@@ -21,12 +89,77 @@ kubectl delete -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-
+   }
+   ```
+
+In the above configuration `vsanfs-sh.prv` is the DNS suffixes for vSAN file Service and `10.161.191.241` is the DNS server that helps resolve file share hostname.
````
`suffixes` -> `suffix` in both sentences.

divyenpatel replied:
done.

shalini-b commented on docs/book/driver-deployment/upgrade.md:
````diff
@@ -21,12 +89,77 @@ kubectl delete -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-
+4. Install Driver
+
+   ```bash
+   kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.3.0-rc.5/manifests/vanilla/vsphere-csi-driver.yaml
````
Do we plan to change `v2.3.0-rc.5` to `v2.3.0` later?

divyenpatel replied:
it should be `v2.3.0`. fixed.

chethanv28 commented on docs/book/driver-deployment/upgrade.md:
```diff
@@ -4,14 +4,82 @@
-
 Each release of vSphere CSI driver has a different set of RBAC rules and deployment YAML files depending on the version of the vCenter on which Kubernetes Cluster is deployed.
 
+## If you have RWM volumes backed by vSAN file service deployed using vSphere CSI Driver, Please refer to the following steps before upgrading vSphere CSI Driver
```
`Please` ==> `please`

divyenpatel replied:
done

chethanv28 commented on docs/book/driver-deployment/upgrade.md:
```diff
@@ -4,14 +4,82 @@
 
+## If you have RWM volumes backed by vSAN file service deployed using vSphere CSI Driver, Please refer to the following steps before upgrading vSphere CSI Driver
+
+vSAN file service volumes (Read Write Many vSphere CSI Volumes) become inaccessible after upgrading the CSI driver. Refer to this [issue](https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/1216) for more detail.
```
Let's use `file volumes` or `file share volumes` consistently at all places

divyenpatel replied:
will use "vsan file share volumes".

chethanv28 commented on docs/book/driver-deployment/upgrade.md:
````diff
@@ -4,14 +4,82 @@
+    10.83.28.38:/52d7e15c-d282-3bae-f64d-8851ad9d352c on /var/lib/kubelet/pods/43686ba4-d765-4378-807a-74049fca39ee/volumes/kubernetes.io~csi/pvc-7e43d1d3-2675-438d-958d-41315f97f42e/mount type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.244.3.134,local_lock=none,addr=10.83.28.38)
+    ```
+
+4. unmount and remount volume at the same location with the same mount options used for mounting volume. For this step, the `nfs-common` package needs to be pre-installed on the worker VMs.
````
`unmount` ==> `Unmount`

divyenpatel replied:
fixed.

chethanv28 commented on docs/book/driver-deployment/upgrade.md:
````diff
@@ -4,14 +4,82 @@
+         root@k8s-worker3:~# umount -f /var/lib/kubelet/pods/43686ba4-d765-4378-807a-74049fca39ee/volumes/kubernetes.io~csi/pvc-7e43d1d3-2675-438d-958d-41315f97f42e/mount
+         ```
+
+      - If the volume is busy or already inaccessible, use `umount -l` command to do the lazy unmount.
````
Instead of 2 instructions for 2 scenarios, can we just include everything(-l option & -f option) in 1 instruction

```
root@k8s-worker3:~# umount -lf /var/lib/kubelet/pods/43686ba4-d765-4378-807a-74049fca39ee/volumes/kubernetes.io~csi/pvc-7e43d1d3-2675-438d-958d-41315f97f42e/mount
```

divyenpatel replied:
Checked with Sriram, we are good with `umoint -fl`. I will make the change.

chethanv28 commented on docs/book/driver-deployment/upgrade.md:
````diff
@@ -4,14 +4,82 @@
+         root@k8s-worker3:~# mount -t nfs4 -o sec=sys,minorversion=1  10.83.28.38:/52d7e15c-d282-3bae-f64d-8851ad9d352c /var/lib/kubelet/pods/43686ba4-d765-4378-807a-74049fca39ee/volumes/kubernetes.io~csi/pvc-7e43d1d3-2675-438d-958d-41315f97f42e/mount
+        ```
+
+5. confirm mount point is accessible from the Node VM
````
`confirm` ==> `Confirm`

divyenpatel replied:
done.

divyenpatel commented:
@chethanv28 @shalini-b @SandeepPissay I have addressed all your review comments. Please take a look at refreshed preview [link](https://deploy-preview-1220--kubernetes-sigs-vsphere-csi-driver.netlify.app/driver-deployment/upgrade.html).

I will also update release notes of v2.0.0 to v2.2.1 releases and link to this issue.

shalini-b commented:
/approve

sashrith commented on docs/book/driver-deployment/upgrade.md:
```diff
@@ -4,14 +4,76 @@
+## If you have RWM volumes backed by vSAN file service deployed using vSphere CSI Driver, please refer to the following steps before upgrading vSphere CSI Driver
+
+vSAN file share volumes (Read Write Many vSphere CSI Volumes) become inaccessible after upgrading the CSI driver. Refer to this [issue](https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/1216) for more detail.
+This issue is present in vSphere CSI Driver v2.0.0 to v2.2.1. Fix for this issue will be provided in `v2.3.0` release. New Pods with file share volumes created using v2.3.0 will not run into this issue.
```
Fixes are being back ported to older releases as well right. This may have to be revised in future, so better to have the exact list here.

divyenpatel replied:
@chethanv28 will be taking care of this when he is releasing patches for older releases.

sashrith commented on docs/book/driver-deployment/upgrade.md:
````diff
@@ -4,14 +4,76 @@
+
+    ```bash
+    # kubectl get pv -o wide | grep RWX
+    pvc-7e43d1d3-2675-438d-958d-41315f97f42e   1Gi        RWX            Delete           Bound    default/www-web-0   file-sc                 107m   Filesystem
````
we can have ReadOnlyMany volumes as well right?

divyenpatel replied:
added.
```
# kubectl get pv -o wide | grep 'RWX\|ROX'
```

sashrith commented on docs/book/driver-deployment/upgrade.md:
```diff
@@ -21,12 +83,77 @@ kubectl delete -f https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-
+    reload
+    loadbalance
+   }
+   vsanfs-sh.prv:53 {
```
We have to add this to installation docs as well right?
https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/a9250f6540080be9868488200b95447d983cea3a/docs/book/driver-deployment/installation.md#install

divyenpatel replied:
we are adding this at https://vsphere-csi-driver.sigs.k8s.io/driver-deployment/prerequisites.html

chethanv28 reviewed (approved):
/approve

chethanv28 commented:
/lgtm
